pkg/grpc/jwt_authenticator.go

```diff
@@ -0,0 +1,137 @@
+package grpc
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"strings"
+
+	"github.com/buildbarn/bb-storage/pkg/clock"
+	"github.com/buildbarn/bb-storage/pkg/util"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+type JWTKeyConfig struct {
+	Key interface{}
+}
+
+type jwtAuthenticator struct {
+	verifyKeys []JWTKeyConfig
+	clock      clock.Clock
+}
+
+// From: https://github.com/square/go-jose/blob/v2/jose-util/utils.go
+func loadJSONWebKey(json []byte, pub bool) (*jose.JSONWebKey, error) {
+	var jwk jose.JSONWebKey
+	err := jwk.UnmarshalJSON(json)
+	if err != nil {
+		return nil, err
+	}
+	// Commenting out due to issue with .Valid returning false on symmetric keys:
+	// https://github.com/square/go-jose/issues/314
```

Comment:
I'd suggest that we don't manually copy-paste code like this. What are your thoughts on just using the documented API for the time being, even if it means symmetric keys are broken? We should work with the go-jose folks to get that fixed.

If you want to use symmetric keys yourself, you can always apply a local change to the go-jose library to do what you want. (That said, consider using asymmetric keys! It's a lot safer!)

Reply:
I'm not sure when they are going to fix it. (I reported the issue in June.) I'll probably just need to submit a fix myself. However, it looks like development has stalled on it, given that the `master` branch has had all of the code removed. The more fundamental issue is whether to use go-jose at all.

As for vendoring this code, I was hesitant to use the original code because the `jose-utils` package is a command-line utility and appears to not be intended as a library. (Plus I needed to disable the `.Valid` check any way given the bug with symmetric keys.)

What are your thoughts on JWT libraries for Go?

Reply:
I think the only two reasonable libraries to use are:
- https://github.com/dgrijalva/jwt-go
- https://github.com/square/go-jose
Both of them don't seem to be maintained that well. That's unfortunate.

I'm fine with adding a local fix for go-jose to our repository if you want. We can let Bazel patch up the library after downloading it. We already do that for some other libraries (see patches/). Alternatively, would dgrijalva's library work out of the box?

Reply:
I added a small patch to fix.